torchao/csrc/cuda/rowwise_scaled_linear_sparse_cutlass/rowwise_scaled_linear_sparse_cutlass.cuh

```diff
@@ -0,0 +1,523 @@
+#pragma once
+
+#include <algorithm>
+#include <optional>
+
+#include <ATen/ATen.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/cuda/CUDAUtils.h>
+#include <c10/util/Exception.h>
+
+#if defined(TORCHAO_USE_CUTLASS) && !defined(_WIN32) &&                   \
+    defined(CUDA_VERSION) && (CUDA_VERSION >= 12020)
+#define BUILD_ROWWISE_SCALED_LINEAR_SPARSE_CUTLASS
+#endif
+
+#if defined(BUILD_ROWWISE_SCALED_LINEAR_SPARSE_CUTLASS)
+#include <cuda_runtime.h>
+#include <cutlass/cutlass.h>
+#include <cutlass/functional.h>
+#include <cutlass/epilogue/collective/default_epilogue.hpp>
+#include <cutlass/epilogue/collective/collective_builder.hpp>
+#include <cutlass/gemm/dispatch_policy.hpp>
+#include <cutlass/gemm/collective/collective_builder.hpp>
+#include <cutlass/gemm/device/gemm_universal_adapter.h>
+#include <cutlass/util/packed_stride.hpp>
+
+#include "cutlass_extensions/common.h"
+#endif
+
+#define OPERATOR_NAME "rowwise_scaled_linear_sparse_cutlass"
+
+namespace torchao {
+
+#if defined(BUILD_ROWWISE_SCALED_LINEAR_SPARSE_CUTLASS)
+template<
+    typename DtypeXq,
+    typename DtypeWq,
+    typename DtypeY,
+    typename UseBias,
+    typename DtypeBias,
+    typename DtypeXScale,
+    typename DtypeWScale,
+    typename TileShape,
+    typename ClusterShape>
+void rowwise_scaled_linear_sparse_kernel_cutlass_sm9x(
+  const at::Tensor& Xq, const at::Tensor& X_scale, const at::Tensor& Wq,
+  const at::Tensor& W_meta, const at::Tensor& W_scale, const at::Tensor& bias,
+  at::Tensor& Y) {
+  // For CUTLASS, sparsified tensor must be the first operand, thus
+  // the result will be calculated as:
+  //    ((Wq @ Xq.T) * W_scale * X_scale.T + bias.T).T
+
+  using SmArch = cutlass::arch::Sm90;
+
+  // Use CUTLASS naming conventions for naming datatypes.
+  using ElementA = DtypeWq;
+  using ElementB = DtypeXq;
+  using ElementD = DtypeY;
+  using ElementAScale = DtypeWScale;
+  using ElementBScale = DtypeXScale;
+  using ElementBias = DtypeBias;
+
+  using LayoutTagA = cutlass::layout::RowMajor;
+  using LayoutTagB = cutlass::layout::ColumnMajor;
+  using LayoutTagD = cutlass::layout::ColumnMajor;
+
+  constexpr auto AlignmentA = 128 / cutlass::sizeof_bits<ElementA>::value;
+  constexpr auto AlignmentB = 128 / cutlass::sizeof_bits<ElementB>::value;
+  constexpr auto AlignmentD = 128 / cutlass::sizeof_bits<ElementD>::value;
+
+  // TODO: use different accumulator datatype if inputs are not float.
+  using ElementAccumulator = float;
+  using ElementCompute = float;
+
+  using ProblemShape = cute::Shape<int, int, int, int>;
+
+  // If KernelTmaWarpSpecializedPingpong used for kernel schedule, the
+  // performance is really bad; on the other side, using
+  // KernelTmaWarpSpecializedPingpongFP8FastAccum doesn't seem to
+  // affect the precision much - thus, sticking with it.
+  using KernelSchedule =
```

Comment:
How much of a performance degredation is slow accumulation? Is it about the same 2x we see for the non-[sparse](https://github.com/pytorch/ao/pull/1325/files#r1868193786) version? 

We have some recipes where this makes a difference for the final accuracy.

Reply:
For Llama generate.py, it's indeed close to 2x; for the benchmark (`benchmarks/benchmark_rowwise_scaled_linear_sparse_cutlass.py`) of the operator only, it's even worse - up to 3x for some shapes.

I was thinking about making `use_fast_accum` an optional argument for the operator.  However, that would mean eventually having this as a template argument, with two options, that would double number of templates instantiated during the compilation, that would slow down the compilation twice.  So I think it's better to keep it as is for now, it could be added if a need arise.  (For the same reason, the number of options for scales/bias/output data types in `.cu` file is kept to minimum - this is stated explicitly in a comment in the `.cuh` file.).